The CheckPaddingOverflow function missed checking some padding values

sallenkey-wei opened the pull request:
In pad node CheckPaddingOverflow function op_context->dims refers to the input tensor's dimensions. The shape of paddings_data should be op_context->dims * 2.

sallenkey-wei committed:
The CheckPaddingOverflow function missed checking some padding values

Signed-off-by: sallenkey-wei <weiyuyinn@gmail.com>

mihaimaruseac reviewed (changes requested):
Can you also add a test, please?

sallenkey-wei commented:
Sorry for the late reply. 

Just move the overflow item in the Int64PaddingOverflow check to the last four pad values to trigger the check failure. Currently, the code can only check the first half of the values.

Should I add the following modifications to pad_test.cc to this pull request？

1. pad_test diff:
diff --git a/tensorflow/lite/kernels/pad_test.cc b/tensorflow/lite/kernels/pad_test.cc
index 6fc7e79719a..3471c1f2aaa 100644
--- a/tensorflow/lite/kernels/pad_test.cc
+++ b/tensorflow/lite/kernels/pad_test.cc
@@ -238,7 +238,7 @@ TEST_F(PadOpTest, Int64PaddingInvalidPadValue) { InvalidPadValue<int64_t>(); }
 TEST_F(PadOpTest, Int64PaddingOverflow) {
   EXPECT_DEATH(PadOpConstModel<int64_t>(
                    {TensorType_FLOAT32, {1, 1, 2, 1}}, {4, 2},
-                   {std::numeric_limits<int64_t>::min(), 0, 1, -1, 2, -1, 0, 0},
+                   {0, 0, 1, -1, 2, -1, std::numeric_limits<int64_t>::min(), 0},
                    {TensorType_FLOAT32}),
                "INT64 padding overflow. Only support value between INT32_MIN "
                "and INT32_MAX.");
  DEATH   ] 2024-11-30 15:31:50.021174: F tensorflow/lite/kernels/test_util.cc:291] Check failed: interpreter_->AllocateTensors() == kTfLiteOk Cannot allocate tensors
[  DEATH   ] 
[  FAILED  ] PadOpTest.Int64PaddingOverflow (141 ms)

2. Output after modification:
[ RUN      ] PadOpTest.Int64PaddingOverflow
tensorflow/lite/kernels/pad_test.cc:244: Failure
Death test: PadOpConstModel<int64_t>( {TensorType_FLOAT32, {1, 1, 2, 1}}, {4, 2}, {0, 0, 1, -1, 2, -1, std::numeric_limits<int64_t>::min(), 0}, {TensorType_FLOAT32})
    Result: died but not with expected error.
  Expected: contains regular expression "INT64 padding overflow. Only support value between INT32_MIN and INT32_MAX."
Actual msg:
[  DEATH   ] ERROR: tensorflow/lite/kernels/pad.cc Pad value has to be greater than equal to 0.
[  DEATH   ] ERROR: Node number 0 (PAD) failed to prepare.
[  DEATH   ] 2024-11-30 15:31:50.021174: F tensorflow/lite/kernels/test_util.cc:291] Check failed: interpreter_->AllocateTensors() == kTfLiteOk Cannot allocate tensors
[  DEATH   ] 
[  FAILED  ] PadOpTest.Int64PaddingOverflow (141 ms)

mihaimaruseac commented:
I think creating a new test that is the old one but with the diff applied is better.

sallenkey-wei commented:
I have added a new test item and changed the termination condition of the for loop to 'padding s_total', as using 'op_comtext ->dims * 2' would result in the following error:

tensorflow/lite/kernels/pad_test.cc:215: Failure
Death test: PadOpConstModel<padding_integer_type>( {TensorType_FLOAT32, {1, 1, 2, 1}}, {3, 2}, {1, 1, 2, 2, 3, 3}, {TensorType_FLOAT32})
    Result: died but not with expected error.
  Expected: contains regular expression "3 != 4"
Actual msg:
[  DEATH   ] ERROR: tensorflow/lite/kernels/pad.cc INT64 padding overflow. Only support value between INT32_MIN and INT32_MAX.
[  DEATH   ] ERROR: Node number 0 (PAD) failed to prepare.
[  DEATH   ] 2024-12-02 00:14:54.891651: F tensorflow/lite/kernels/test_util.cc:291] Check failed: interpreter_->AllocateTensors() == kTfLiteOk Cannot allocate tensors
[  DEATH   ] 
[  FAILED  ] PadOpTest.Int64PaddingUnequalDimensions (146 ms)

tensorflow/lite/kernels/pad_test.cc:647: Failure
Death test: f({TensorType_FLOAT32, {1, 1, 2, 1}}, {3, 2}, {1, 1, 2, 2, 3, 3}, 0.0, {TensorType_FLOAT32})
    Result: died but not with expected error.
  Expected: contains regular expression "3 != 4"
Actual msg:
[  DEATH   ] ERROR: tensorflow/lite/kernels/pad.cc INT64 padding overflow. Only support value between INT32_MIN and INT32_MAX.
[  DEATH   ] ERROR: Node number 0 (PADV2) failed to prepare.
[  DEATH   ] 2024-12-02 00:14:56.529107: F tensorflow/lite/kernels/test_util.cc:291] Check failed: interpreter_->AllocateTensors() == kTfLiteOk Cannot allocate tensors
[  DEATH   ] 
[  FAILED  ] PadV2OpTest.Int64PaddingUnequalDimensions (145 ms)

sallenkey-wei commented:
Hi, I saw that one of the CI test items failed. Is it due to a code conflict? Do I need to update the code again

mihaimaruseac commented:
Hi. I cannot check for the next few weeks, but I'll try to look as soon as I get a chance.
